### PR TITLE
Disjoint sampling

### PR DESCRIPTION
**Pull request by RexYing**
The main goal is to allow disjoint sampling (mostly used for temporal sampling use cases), in order to avoid the same node that are sampled multiple times have its timestamps overriding each other.

This is only a tentative prototype, since I find it somewhat difficult to directly integrate this feature into the c++ sampling routine in torch_sparse. But I'm happy to discuss alternatives and pros / cons and potentially revise it significantly.

I'm also yet to do the collation (mentioned in TODO), but it's not hard in python since we have existing routines. It's more difficult if I were to write this in c++.

One disadvantage of this workaround is that the sampling logic is somewhat exposed to the python side in PyG, which could be confusing. There might also be efficiency concerns, since we now loop it over in Python, rather than in c++.

### CODE REVIEW

File: torch_geometric/loader/neighbor_loader.py [hunk 1]
```diff
@@ -198,8 +205,16 @@ def __call__(self, index: Union[List[int], Tensor]):
...
         elif self.data_cls == 'custom' or issubclass(self.data_cls,
                                                      HeteroData):
-            return self._hetero_sparse_neighbor_sample(
-                {self.input_type: index}) + (index.numel(), )
+            if not self.disjoint:
+                return self._hetero_sparse_neighbor_sample(
+                    {self.input_type: index}) + (index.numel(), )
+            else:
+                result = []
+                for i in index.size(0):
```
Comment (Padarn) [inline torch_geometric/loader/neighbor_loader.py:213]: Does this seem like it would be very slow?

File: torch_geometric/loader/utils.py [hunk 2]
```diff
@@ -187,6 +187,31 @@ def filter_hetero_data(
     return out
 
 
+def filter_hetero_data_disjoint(
```
Comment (Padarn) [inline torch_geometric/loader/utils.py:190]: Not used yet?

**Comment by Padarn** [comment]
@RexYing can you help me understand the problem a bit more? Does this have any impact in the case of graphs without a temporal component?

**Comment by rusty1s** [comment]
@Padarn This will have no real effect on graphs without temporal sampling. It will create disjoint/isolated graphs during sampling instead of merging nodes that appear in different subgraphs (which we currently do). This scenario is likely not happening too frequently anyway for very large graphs.

**Comment by Padarn** [comment]
Yep makes sense 👍 This is what I understood from the current implementation.

File: torch_geometric/loader/neighbor_loader.py (hunk 1 above)
Comment (RexYing) [reply torch_geometric/loader/neighbor_loader.py:213]: > Yep makes sense 👍 This is what I understood from the current implementation.

Thanks for looking into it! I'll make the suggested change you made:)

I'm testing out both approaches of doing it in PyG v.s. c++ implementation in torch sparse (and later pyg-lib) to see which one could be better. Yes it does seem slower when doing it in PyG. However, this approach of separate calls to c++ could potentially benefit from easy parallelization (which is currently not utilized here).
So I'll get back to you with more insights once I can test both options out more thoroughly.

**Comment by RexYing** [comment]
> Yep makes sense 👍 This is what I understood from the current implementation.

Thanks for looking into it! I'll make the suggested change you made:)

I'm testing out both approaches of doing it in PyG v.s. c++ implementation in torch sparse (and later pyg-lib) to see which one could be better. Yes it does seem slower when doing it in PyG. However, this approach of separate calls to c++ could potentially benefit from easy parallelization (which is currently not utilized here).
So I'll get back to you with more insights once I can test both options out more thoroughly.

**Comment by rusty1s** [comment]
Closing this since this functionality has been directly integrated into `torch-sparse`.

**Pull request closed without merging**